### PR TITLE
[ADP-3212] Store and report voting state

### PR DESCRIPTION
**Pull request by paolino**
Mildly breaking change to the API (new information added)

- [x] Implement DB persistence of voting
- [x] Update reporting delegation state in the API

ADP-3212

### CODE REVIEW

File: lib/wallet/api/http/Cardano/Wallet/Api/Types.hs [hunk 1]
```diff
@@ -1000,6 +1000,7 @@ data ApiWalletDelegation = ApiWalletDelegation
 data ApiWalletDelegationNext = ApiWalletDelegationNext
     { status :: !ApiWalletDelegationStatus
     , target :: !(Maybe (ApiT PoolId))
+    , voted :: !(Maybe (ApiT VoteAction))
```
Comment (paweljakubas) [inline lib/wallet/api/http/Cardano/Wallet/Api/Types.hs]:
```suggestion
    , voted :: !(Maybe (ApiT DRep))
```

File: lib/wallet/api/http/Cardano/Wallet/Api/Types/Certificate.hs [hunk 2]
```diff
@@ -286,4 +325,50 @@ mkApiAnyCertificate acct' acctPath' = \case
...
+            CastVoteExternal (ApiRewardAccount rewardKey) (ApiT vote')
+    toApiDelCert acctM acctPath (W.CertDelegateAndVoteFull rewardKey poolId' vote') =
+        if Just rewardKey == acctM then
+            WalletDelegationCertificate $
+            JoinPoolCastVote (NE.map ApiT acctPath) (ApiT poolId') (ApiT vote')
+        else
+            DelegationCertificate $
+            JoinPoolCastVoteExternal (ApiRewardAccount rewardKey) (ApiT poolId') (ApiT vote')
+
+instance ToJSON (ApiT DRep) where
```
Comment (paweljakubas) [inline lib/wallet/api/http/Cardano/Wallet/Api/Types/Certificate.hs:345]: probably redundant

File: lib/wallet/api/http/Cardano/Wallet/Api/Types/Certificate.hs [hunk 3]
```diff
@@ -286,4 +325,50 @@ mkApiAnyCertificate acct' acctPath' = \case
...
+    toJSON (ApiT Abstain) = "abstain"
+    toJSON (ApiT NoConfidence) = "no_confidence"
+    toJSON (ApiT (FromDRepID drep)) = case drep of
+        DRepFromKeyHash keyhash ->
+            String $ encodeDRepKeyHashBech32 keyhash
+        DRepFromScriptHash scripthash ->
+            String $ encodeDRepScriptHashBech32 scripthash
+instance FromJSON (ApiT DRep) where
```
Comment (paweljakubas) [inline lib/wallet/api/http/Cardano/Wallet/Api/Types/Certificate.hs:353]: probably redundant

File: lib/wallet/test/data/Cardano/Wallet/Api/ApiTVoteAction.json [hunk 4]
```diff
@@ -0,0 +1,15 @@
+{
```
Comment (paweljakubas) [inline lib/wallet/test/data/Cardano/Wallet/Api/ApiTVoteAction.json]: this golden not needs, already present in master

File: lib/wallet/test/unit/Cardano/Wallet/DB/Arbitrary.hs [hunk 5]
```diff
@@ -157,6 +157,12 @@ import Cardano.Wallet.Primitive.Types.Address
 import Cardano.Wallet.Primitive.Types.Coin
     ( Coin (..)
     )
+import Cardano.Wallet.Primitive.Types.DRep
+    ( DRep (..)
+    , DRepKeyHash (..)
+    , DRepScriptHash (..)
+    , VoteAction (..)
```
Comment (paweljakubas) [inline lib/wallet/test/unit/Cardano/Wallet/DB/Arbitrary.hs]:
```suggestion
    , DRep (..)
```

File: lib/wallet/test/unit/Cardano/Wallet/DB/Arbitrary.hs [hunk 6]
```diff
@@ -157,6 +157,12 @@ import Cardano.Wallet.Primitive.Types.Address
 import Cardano.Wallet.Primitive.Types.Coin
     ( Coin (..)
     )
+import Cardano.Wallet.Primitive.Types.DRep
+    ( DRep (..)
```
Comment (paweljakubas) [inline lib/wallet/test/unit/Cardano/Wallet/DB/Arbitrary.hs:161]:
```suggestion
    ( DRepID (..)
```

File: lib/wallet/test/unit/Cardano/Wallet/DB/Arbitrary.hs [hunk 7]
```diff
@@ -935,6 +941,17 @@ instance Arbitrary ChangeAddressMode where
     arbitrary = genericArbitrary
     shrink = genericShrink
 
+instance Arbitrary DRep where
```
Comment (paweljakubas) [inline lib/wallet/test/unit/Cardano/Wallet/DB/Arbitrary.hs]:
```suggestion
instance Arbitrary DRepID where
```

File: lib/wallet/test/unit/Cardano/Wallet/DB/Arbitrary.hs [hunk 8]
```diff
@@ -935,6 +941,17 @@ instance Arbitrary ChangeAddressMode where
...
+    arbitrary = do
+        InfiniteList bytes _ <- arbitrary
+        oneof [ pure $ DRepFromKeyHash $ DRepKeyHash $ BS.pack $ take 28 bytes
+              , pure $ DRepFromScriptHash $ DRepScriptHash $ BS.pack $ take 28 bytes
+              ]
+
+instance Arbitrary VoteAction where
```
Comment (paweljakubas) [inline lib/wallet/test/unit/Cardano/Wallet/DB/Arbitrary.hs]:
```suggestion
instance Arbitrary DRep where
```

File: lib/wallet/test/unit/Cardano/Wallet/DB/Arbitrary.hs [hunk 9]
```diff
@@ -935,6 +941,17 @@ instance Arbitrary ChangeAddressMode where
...
+    arbitrary =
+        oneof [pure Abstain, pure NoConfidence, VoteTo <$> arbitrary]
```
Comment (paweljakubas) [inline lib/wallet/test/unit/Cardano/Wallet/DB/Arbitrary.hs]: not `VoteTo` anymore in new Drep definition

File: lib/wallet/test/unit/Cardano/Wallet/DelegationSpec.hs [hunk 10]
```diff
@@ -34,6 +34,12 @@ import Cardano.Wallet.Primitive.Types.Coin
 import Cardano.Wallet.Primitive.Types.Coin.Gen
     ( genCoinPositive
     )
+import Cardano.Wallet.Primitive.Types.DRep
+    ( DRep (..)
```
Comment (paweljakubas) [inline lib/wallet/test/unit/Cardano/Wallet/DelegationSpec.hs:38]:
```suggestion
    ( DRepID (..)
```

File: lib/wallet/test/unit/Cardano/Wallet/DelegationSpec.hs [hunk 11]
```diff
@@ -34,6 +34,12 @@ import Cardano.Wallet.Primitive.Types.Coin
...
+    , DRepKeyHash (..)
+    , DRepScriptHash (..)
+    , VoteAction (..)
```
Comment (paweljakubas) [inline lib/wallet/test/unit/Cardano/Wallet/DelegationSpec.hs]:
```suggestion
    , DRep (..)
```

File: lib/wallet/test/unit/Cardano/Wallet/DelegationSpec.hs [hunk 12]
```diff
@@ -288,3 +295,14 @@ instance Arbitrary Word31 where
 instance Arbitrary a => Arbitrary (NonEmpty a) where
     arbitrary = (:|) <$> arbitrary <*> arbitrary
     shrink = genericShrink
+
+instance Arbitrary DRep where
```
Comment (paweljakubas) [inline lib/wallet/test/unit/Cardano/Wallet/DelegationSpec.hs]:
```suggestion
instance Arbitrary DRepID where
```

File: lib/wallet/test/unit/Cardano/Wallet/DelegationSpec.hs [hunk 13]
```diff
@@ -288,3 +295,14 @@ instance Arbitrary Word31 where
...
+    arbitrary = do
+        InfiniteList bytes _ <- arbitrary
+        oneof [ pure $ DRepFromKeyHash $ DRepKeyHash $ BS.pack $ take 28 bytes
+              , pure $ DRepFromScriptHash $ DRepScriptHash $ BS.pack $ take 28 bytes
+              ]
+
+instance Arbitrary VoteAction where
```
Comment (paweljakubas) [inline lib/wallet/test/unit/Cardano/Wallet/DelegationSpec.hs]:
```suggestion
instance Arbitrary DRep where
```

File: lib/wallet/test/unit/Cardano/Wallet/DelegationSpec.hs [hunk 14]
```diff
@@ -288,3 +295,14 @@ instance Arbitrary Word31 where
...
+    arbitrary =
+        oneof [pure Abstain, pure NoConfidence, VoteTo <$> arbitrary]
```
Comment (paweljakubas) [inline lib/wallet/test/unit/Cardano/Wallet/DelegationSpec.hs]: VoteTo was changed

File: specifications/api/swagger.yaml [hunk 15]
```diff
@@ -393,6 +393,22 @@ x-drepScriptHash: &drepScriptHash
   description: DRep's script hash.
   pattern: "^(drep_script)1[0-9a-z]*$"
 
+x-noVote: &noVote
```
Comment (paweljakubas) [inline specifications/api/swagger.yaml]: noVote and anyVoting NOT needed anymore

File: specifications/api/swagger.yaml [hunk 16]
```diff
@@ -2908,6 +2926,14 @@ components:
       description: |
         The result of signing a transaction (serialized and encoded).
 
+    ApiVoteAction: &ApiVoteAction
```
Comment (paweljakubas) [inline specifications/api/swagger.yaml]: ApiVoteAction change NOT needed anymore

**Comment by paweljakubas** [comment]
there was also commit changing Delegations.md that should be added

**Review by paweljakubas — APPROVED** [review]
should be fine

File: specifications/api/swagger.yaml [hunk 17]
```diff
@@ -2917,6 +2938,7 @@ components:
       properties:
         status: *delegationStatus
         target: *delegationTarget
+        voted: *ApiDRep
```
Comment (HeinrichApfelmus) [inline specifications/api/swagger.yaml]:
```suggestion
        voting: *ApiDRep
```

File: specifications/api/swagger.yaml [hunk 18]
```diff
@@ -2939,6 +2961,7 @@ components:
           properties:
             status: *delegationStatus
             target: *delegationTarget
+            voted: *ApiDRep
```
Comment (HeinrichApfelmus) [inline specifications/api/swagger.yaml]:
```suggestion
            voting: *ApiDRep
```

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs [hunk 19]
```diff
@@ -1233,12 +1233,28 @@ toApiWalletDelegation W.WalletDelegation{active,next} ti = do
         W.Delegating pid -> ApiWalletDelegationNext
             { status = Delegating
             , target = Just (ApiT pid)
+            , voted = Nothing
```
Comment (HeinrichApfelmus) [inline lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs]:
```suggestion
            , voting = Nothing
```

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs [hunk 20]
```diff
@@ -1233,12 +1233,28 @@ toApiWalletDelegation W.WalletDelegation{active,next} ti = do
...
             , changesAt = mepochInfo
             }
 
         W.NotDelegating -> ApiWalletDelegationNext
             { status = NotDelegating
             , target = Nothing
+            , voted = Nothing
```
Comment (HeinrichApfelmus) [inline lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs]:
```suggestion
            , voting = Nothing
```

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs [hunk 21]
```diff
@@ -1233,12 +1233,28 @@ toApiWalletDelegation W.WalletDelegation{active,next} ti = do
...
+            , changesAt = mepochInfo
+            }
+
+        W.Voting voting -> ApiWalletDelegationNext
+            { status = Voting
+            , target = Nothing
+            , voted =  Just (ApiT voting)
```
Comment (HeinrichApfelmus) [inline lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs]:
```suggestion
            , voting =  Just (ApiT voting)
```

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs [hunk 22]
```diff
@@ -1233,12 +1233,28 @@ toApiWalletDelegation W.WalletDelegation{active,next} ti = do
...
+            , changesAt = mepochInfo
+            }
+
+        W.DelegatingVoting pid voting -> ApiWalletDelegationNext
+            { status = DelegatingAndVoting
+            , target = Just (ApiT pid)
+            , voted =  Just (ApiT voting)
```
Comment (HeinrichApfelmus) [inline lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs]:
```suggestion
            , voting =  Just (ApiT voting)
```

File: lib/wallet/api/http/Cardano/Wallet/Api/Types.hs [hunk 23]
```diff
@@ -1000,6 +1000,7 @@ data ApiWalletDelegation = ApiWalletDelegation
 data ApiWalletDelegationNext = ApiWalletDelegationNext
     { status :: !ApiWalletDelegationStatus
     , target :: !(Maybe (ApiT PoolId))
+    , voted :: !(Maybe (ApiT DRep))
```
Comment (HeinrichApfelmus) [inline lib/wallet/api/http/Cardano/Wallet/Api/Types.hs]:
```suggestion
    , voting :: !(Maybe (ApiT DRep))
```

File: lib/wallet/src/Cardano/Wallet/DB/Pure/Implementation.hs
Comment (HeinrichApfelmus) [inline lib/wallet/src/Cardano/Wallet/DB/Pure/Implementation.hs:1]: We no longer use the State Machine to test our database, so we can leave this out.

File: lib/wallet/src/Cardano/Wallet/DB/Store/Delegations/Layer.hs [hunk 24]
```diff
@@ -60,6 +61,16 @@ isStakeKeyRegistered m = fromMaybe False $ do
     (_, v) <- lookupMax m
     pure $ v /= Inactive
 
+-- | Check whether the voting has been casted in the delegation state.
+hasAlreadyVoted :: Delegations -> Bool
+hasAlreadyVoted m = fromMaybe False $ do
```
Comment (HeinrichApfelmus) [inline lib/wallet/src/Cardano/Wallet/DB/Store/Delegations/Layer.hs]:
```suggestion
-- | Check whether the vote has been delegated to a 'DRep'.
isVoting :: Delegations -> Bool
isVoting m = fromMaybe False $ do
```

File: lib/wallet/src/Cardano/Wallet/DB/Store/Delegations/Store.hs [hunk 25]
```diff
@@ -83,17 +83,27 @@ loadS' = do
 data DecodeDelegationError
     = ActiveDelegationWithoutAPool
     | UnknownDelegationStatus Int
+    | VotingWithoutAVoteAction
+    | ActiveVotingWithoutAVoteAction
+    | ActiveVotingWithoutAPool
+    | ActiveVotingWithoutPoolAndVoteAction
```
Comment (HeinrichApfelmus) [inline lib/wallet/src/Cardano/Wallet/DB/Store/Delegations/Store.hs]:
```suggestion
    | VotingWithoutADRep
    | ActiveVotingWithoutADRep
    | ActiveVotingWithoutAPool
    | ActiveVotingWithoutPoolAndDRep
```

File: lib/wallet/src/Cardano/Wallet/Delegation/Model.hs [hunk 26]
```diff
@@ -32,6 +35,8 @@ data Operation slot pool
     = Register slot
     | Deregister slot
     | Delegate pool slot
+    | DelegateAndVote pool DRep slot
+    | Vote DRep slot
```
Comment (HeinrichApfelmus) [inline lib/wallet/src/Cardano/Wallet/Delegation/Model.hs:39]:
```suggestion
    = VoteAndDelegate (Maybe DRep) (Maybe pool) slot
    | Deregister slot
```

This would follow the Conway specification more closely:
conway DCert

As this constructor from the specification puts the voting argument first, so I have called it `VoteAndDelegate` rather than `DelegateAndVote`.

File: lib/wallet/src/Cardano/Wallet/Delegation/Model.hs [hunk 27]
```diff
@@ -40,10 +45,17 @@ slotOf :: Operation slot pool -> slot
...
 slotOf (Rollback x) = x
 
 -- | Valid state for the delegations, independent of time.
-data Status pool = Inactive | Registered | Active pool
+data Status pool
+    = Inactive
+    | Registered
+    | Active pool
+    | ActiveAndVoted pool DRep
+    | Voted DRep
```
Comment (HeinrichApfelmus) [inline lib/wallet/src/Cardano/Wallet/Delegation/Model.hs:58]:
```suggestion
data Status pool
    = Inactive
    | Active (Maybe DRep) (Maybe pool)

pattern Registered = Active Nothing Nothing
pattern Delegating pool = Active Nothing (Just pool)
pattern Voting drep = Active (Just drep) Nothing
pattern VotingAndDelegating pool drep = Active (Just drep) (Just pool)
```

As discussed: Simpler internal implementation while keeping the patterns.

File: lib/wallet/api/http/Cardano/Wallet/Api/Types.hs [hunk 28]
```diff
@@ -1009,6 +1010,8 @@ data ApiWalletDelegationNext = ApiWalletDelegationNext
 data ApiWalletDelegationStatus
     = NotDelegating
     | Delegating
+    | Voting
+    | DelegatingAndVoting
```
Comment (HeinrichApfelmus) [inline lib/wallet/api/http/Cardano/Wallet/Api/Types.hs]:
```suggestion
    | VotingAndDelegating
```

See below, to keep in line with "voting argument first" from the Conway spec.

File: lib/wallet/test/unit/Cardano/Wallet/DB/Store/Delegations/StoreSpec.hs
Comment (HeinrichApfelmus) [inline lib/wallet/test/unit/Cardano/Wallet/DB/Store/Delegations/StoreSpec.hs:1]: The unit tests are nice, but in order to get substantial coverage, we should update the `genDelta` function as well, or rather the `Config` type. (I think that it's ok to do that in a separate pull request, though.)

**Review by HeinrichApfelmus — APPROVED** [review]
Thanks! 😊 

I have several requests on name changes / constructor rearrangement, but the rest looks good to me.

Also, we probably want to create more deltas in the `genDelta` function in order to properly test the `Store` laws. However, I'm fine with doing that in a separate pull request.

File: lib/wallet/src/Cardano/Wallet/Delegation/Model.hs (hunk 26 above)
Comment (paolino) [reply lib/wallet/src/Cardano/Wallet/Delegation/Model.hs:39]: will do in a separate PR

File: lib/wallet/src/Cardano/Wallet/Delegation/Model.hs (hunk 27 above)
Comment (paolino) [reply lib/wallet/src/Cardano/Wallet/Delegation/Model.hs:58]: same as previous one

File: lib/wallet/test/unit/Cardano/Wallet/DB/Store/Delegations/StoreSpec.hs
Comment (paolino) [reply lib/wallet/test/unit/Cardano/Wallet/DB/Store/Delegations/StoreSpec.hs:1]: another PR for this
